src/core/ipe-setup.c

```diff
@@ -0,0 +1,111 @@
+                const char *suffix;
+
+                r = path_extract_filename(*policy, &file_name);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to extract filename from IPE policy path %s: %m", *policy);
+
+                /* Filtered by conf_files_list_nulstr() */
+                suffix = ASSERT_PTR(endswith(file_name, ".p7b"));
+
+                policy_name = strndup(file_name, suffix - file_name);
+                if (!policy_name)
+                        return log_oom();
+
+                if (!filename_is_valid(policy_name))
+                        return log_error_errno(
+                                        SYNTHETIC_ERRNO(EINVAL),
+                                        "Invalid IPE policy name %s",
+                                        policy_name);
+
+                input = open(*policy, O_RDONLY|O_NOFOLLOW|O_CLOEXEC);
+                if (input < 0)
+                        return log_error_errno(
+                                        errno,
+                                        "Failed to open the IPE policy file %s: %m",
+                                        *policy);
+
+                /* If policy is already installed, try to update it */
+                output_path = path_join(IPE_SECFS_POLICIES, policy_name, "update");
+                if (!output_path)
+                        return log_oom();
+
+                output = open(output_path, O_WRONLY|O_CLOEXEC);
+                if (output < 0 && errno == ENOENT)
+                        /* Policy is not installed, install it and activate it */
+                        output = open(IPE_SECFS_NEW_POLICY, O_WRONLY|O_CLOEXEC);
+                if (output < 0)
+                        return log_error_errno(
+                                        errno,
+                                        "Failed to open the IPE policy handle for writing: %m");
+
+                /* The policy is inline signed in binary format, so it has to be copied in one go, otherwise the
+                 * kernel will reject partial inputs with -EBADMSG. */
+                r = copy_bytes(input, output, UINT64_MAX, /* copy_flags= */ 0);
+                if (r < 0)
+                        return log_error_errno(
+                                        r,
+                                        "Failed to copy the IPE policy %s to %s: %m",
+                                        *policy,
+                                        output_path);
+
+                output = safe_close(output);
+
+                activate_path = path_join(IPE_SECFS_POLICIES, policy_name, "active");
+                if (!activate_path)
+                        return log_oom();
+
+                r = write_string_file(activate_path, "1", 0);
+                if (r == -ESTALE) {
+                        log_debug("IPE policy %s is already loaded with a version that is equal or higher, skipping.",
```

Comment:
Let's still embed the errno in structured log even though it's not logged, i.e. log_debug_errno